packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -1484,32 +1502,63 @@ export default function makeKernelKeeper(kernelStorage, kernelSlog) {
             deleteKernelPromise(kpid);
           }
         }
+
         if (type === 'object') {
           const { reachable, recognizable } = getObjectRefCount(kref);
           if (reachable === 0) {
-            const ownerVatID = ownerOfKernelObject(kref);
-            if (ownerVatID) {
+            // We avoid ownerOfKernelObject(), which will report
+            // 'undefined' if the owner is dead (and being slowly
+            // deleted). Message delivery should use that, but not us.
+            const ownerKey = `${kref}.owner`;
+            let ownerVatID = kvStore.get(ownerKey);
+            const terminated = terminatedVats.includes(ownerVatID);
+
+            if (ownerVatID && !terminated) {
               const vatKeeper = provideVatKeeper(ownerVatID);
               const isReachable = vatKeeper.getReachableFlag(kref);
               if (isReachable) {
                 // the reachable count is zero, but the vat doesn't realize it
                 actions.add(`${ownerVatID} dropExport ${kref}`);
               }
               if (recognizable === 0) {
-                // TODO: rethink this
+                // TODO: rethink this assert
                 // assert.equal(isReachable, false, `${kref} is reachable but not recognizable`);
                 actions.add(`${ownerVatID} retireExport ${kref}`);
               }
-            } else if (recognizable === 0) {
-              // unreachable, unrecognizable, orphaned: delete the
-              // empty refcount here, since we can't send a GC
-              // action without an ownerVatID
-              deleteKernelObject(kref);
+            }
+            if (ownerVatID && terminated) {
```

Comment:
```suggestion
            } else if (ownerVatID && terminated) {
```
?

Reply:
Yeah, those first two clauses are mutually-exclusive (1: `ownerID && !terminated`, 2: `ownerID && terminated`). But in applying your suggested change (which would improve things by making the mutual-exclusivity more obvious), I ignored my own comments and tried to make all three clauses mututally exclusive (changing the `if (!ownerVatID)` a dozen lines down into an `else if`), and *that* would be wrong. It's important that we fall through to the orphaned case when we've stopped recognizing the unreachable+orphaned reference.

My near-mistake suggests that this code is a bit too tricky.

I'm not seeing an obvious way to refactor this to improve that, so I think I'll apply your change, but also add a bigger comment at the end, to discourage folks from attempting to factor out the `ownerVatID` clause in the future.